intel-hex.js

```diff
@@ -817,6 +817,40 @@ class MemoryMap {
         return memMap;
     }
 
+
+    /**
+     * Returns a new instance of <tt>MemoryMap</tt>, containing only data between
+     * <tt>offset</tt> and <tt>offset</tt> + <tt>length</tt>.
+     *
+     * <br/>
+     * The returned <tt>MemoryMap</tt> might be empty.
+     * <br/>
+     * Internally, this uses <tt>subarray</tt>, so new memory is not allocated.
+     *
+     * @param {Number} offset The start offset
+     * @param {Number} length The lenght of memory map to slice out
+     * @return {MemoryMap}
+     */
+    slice(offset, length){
+        const sliced = new MemoryMap();
+
+        for (let [blockAddr, block] of this) {
+            const blockLength = block.length;
+
+            if ((blockAddr + length) >= offset && blockAddr < (offset + length)) {
```

Comment:
Yes, basically flipping the if to return if we are outside of the wanted range of condition.

Reply:
I don't think that's possible, because the insertion order of the data blocks is not always guaranteed to be strictly ascending. Specifically:
```
let memMap = new MemoryMap();
memMap.set(0x0000, new Uint8Array([....]));
memMap.set(0x9000, new Uint8Array([....]));
memMap.set(0x8000, new Uint8Array([....]));
let slices = memMap.slice(0x8000, 0x10)
```
In this case, the `for (... of ...)` loop will iterate in the insertion order. Returning early on the 0x0000 or 0x9000 blocks would mean ignoring the third block, which is relevant for the sliced data.

Reply:
Right, my bad, I said *return* but in this case it would be *continue*, to skip to next iteration, that should not be a problem. I don't mean to be bikeshedding, it's also just to highlight the general principle.

Reply:
Yeah, but `continue` is considered to be more prone to mistakes (see e.g. https://eslint.org/docs/rules/no-continue ), and I share that opinion. I'll leave the loop logic as it is now.

Reply:
Ok, good point, agreed.